internal/controller/httproute_controller.go

```diff
@@ -595,28 +590,36 @@ func (r *HTTPRouteReconciler) processHTTPRoute(tctx *provider.TranslateContext,
 	return terror
 }
 
-func (r *HTTPRouteReconciler) httpRoutePolicyPredicateOnUpdate(e event.UpdateEvent) bool {
-	oldPolicy, ok0 := e.ObjectOld.(*v1alpha1.HTTPRoutePolicy)
-	newPolicy, ok1 := e.ObjectNew.(*v1alpha1.HTTPRoutePolicy)
-	if !ok0 || !ok1 {
-		return false
-	}
-	var discardsRefs = make(map[string]v1alpha2.LocalPolicyTargetReferenceWithSectionName)
-	for _, ref := range oldPolicy.Spec.TargetRefs {
-		key := indexer.GenHTTPRoutePolicyIndexKey(string(ref.Group), string(ref.Kind), e.ObjectOld.GetNamespace(), string(ref.Name), "")
-		discardsRefs[key] = ref
-	}
-	for _, ref := range newPolicy.Spec.TargetRefs {
-		key := indexer.GenHTTPRoutePolicyIndexKey(string(ref.Group), string(ref.Kind), e.ObjectOld.GetNamespace(), string(ref.Name), "")
-		delete(discardsRefs, key)
-	}
-	if len(discardsRefs) > 0 {
-		dump := oldPolicy.DeepCopy()
-		dump.Spec.TargetRefs = make([]v1alpha2.LocalPolicyTargetReferenceWithSectionName, 0, len(discardsRefs))
-		for _, ref := range discardsRefs {
-			dump.Spec.TargetRefs = append(dump.Spec.TargetRefs, ref)
+func httpRoutePolicyPredicateOnUpdate(c chan event.GenericEvent, kind string) func(e event.UpdateEvent) bool {
+	return func(e event.UpdateEvent) bool {
+		oldPolicy, ok0 := e.ObjectOld.(*v1alpha1.HTTPRoutePolicy)
+		newPolicy, ok1 := e.ObjectNew.(*v1alpha1.HTTPRoutePolicy)
+		if !ok0 || !ok1 {
+			return false
+		}
+		var discardsRefs = make(map[string]v1alpha2.LocalPolicyTargetReferenceWithSectionName)
+		for _, ref := range oldPolicy.Spec.TargetRefs {
+			if string(ref.Kind) != kind {
+				continue
+			}
+			key := indexer.GenHTTPRoutePolicyIndexKey(string(ref.Group), string(ref.Kind), e.ObjectOld.GetNamespace(), string(ref.Name), "")
+			discardsRefs[key] = ref
+		}
+		for _, ref := range newPolicy.Spec.TargetRefs {
+			if string(ref.Kind) != kind {
```

Comment:
I think it would be more appropriate for the consumer of the event to decide on the filtering criteria.
What's your opinion?

Reply:
Your approach has higher cohesion.
After your PR is merged, I will change it to your style.